### DISCUSSION TITLE
A language server for tmux.conf

### DISCUSSION THREAD

**Discussion by Freed-Wu** [tmux/tmux#3780] [General]
Hey, I write [a language server](https://github.com/Freed-Wu/tmux-language-server) to complete tmux command and option. I believe it should be helpful for many people. Can I host it on tmux github organization to make more people know and try it? It can help to get more bug reports to improve it and increase its [bus factor](https://en.wikipedia.org/wiki/Bus_factor). TIA!

A demo video is here: <https://asciinema.org/a/628457>

![628457](https://github.com/tmux/tmux/assets/32936898/9f0e4d45-e286-4191-97de-5c2d60b7c334)

**Comment by nicm** [comment]
What does it do?

**Reply by Freed-Wu** [reply]
> complete tmux command and option.

complete `~/.config/tmux/tmux.conf`.

**Comment by krader1961** [comment]
@nicm, a LSP is a way to augment the behavior of an editor or IDE when a user is editing a particular language (such as the Tmux config file) so that tab completion produces meaningful options. It can also provide other useful features such as jumping to the location where a symbol is defined or give a list of references to a symbol. I use an LSP for the Go language with ViM to make it easier to find where a symbol is defined or referenced. However, I would be wary of merging an LSP implementation for the Tmux confg file into your project. What might be justified is adding a link to https://github.com/Freed-Wu/tmux-language-server. Personally, I can't imagine using it because I rarely edit my *~/.tmux.conf*, and the file is so small (at least compared to a typical program written C++/Go/Java), that I just don't see the utility.

**Reply by Freed-Wu** [reply]
Oh, I think you are a professional tmux user and familiar with tmux's keywords and options :smile: but for new users, a language server will be good.

> the file is so small

And this language server is just a start, some function (like diagnosis of option's type) is missing. And a parser based on [tree-sitter](https://tree-sitter.github.io/tree-sitter/) to provide a better information is future's work. So currently, small is normal.

In fact, It is not the only language server I write, Such as:

- <https://github.com/termux/termux-language-server>: language server for PKGBUILD, ebuild, build.sh
- <https://github.com/neomutt/mutt-language-server>: language server for `~/.config/neomutt/neomuttrc`

I do this work just because there are many Domain Specific Languages like `~/.config/tmux/tmux.conf`. They cannot get good support of language server and edit them is a little hard for many new users. I hope I can reduce their difficulty.

**Reply by nicm** [reply]
We used to link to the bash completion from somewhere but I can't find where now. But we can add a section to the FAQ.

But you will need to be more specific. What editors is it for?

**Reply by Freed-Wu** [reply]
All editors supporting LSP are here: <https://microsoft.github.io/language-server-protocol/implementors/tools/>
